Step3 - PR

yunbaek opened the pull request:
안녕하세요! 백경윤입니다.

ATTD 미션 Step3 - 구간 추가 미션 구현하여 PR 드립니다.

개인 사정이 있어 미션이 늦어졌습니다.
경황이 없어 꼼꼼하게 신경쓰지 못 한 부분이 많은데
이번에도 따끔한 지적 부탁드리겠습니다 😂 

항상 세심한 리뷰 감사드립니다!

Commit:
fix: datasource - statement 닫지 않았던 버그 수정

Commit:
docs: 요구사항 정리

Commit:
docs: 구간 추가 성공 인수 조건 작성

Commit:
test: 구간 추가 성공 케이스에 대한 인수 테스트 작성

Commit:
feat: Section 도메인, 일급컬렉션 생성

- `Line` 에 대해 일대다 설정
- `상행역`, `하행역` 에 대해 일대다 설정
- `@Embeddable` 적용

Commit:
feat: 중복되는 연관관계 제거

- `line` - `station` 직접 연관관계 제거 (`section` 을 통한 `@ManyToMany` 관계 설정)
- station 반환 로직 수정 (기존 상행역, 하행역 반환 -> 등록된 모든 역 반환)
- `Station` equals & hashcode 재정의

Commit:
test: 구간 추가 실패 케이스에 대한 인수 테스트 작성

Commit:
feat: 구간 조회, 추가 관련 기능 추가

- 모든 역 조회
- 상향 종점 조회
- 하향 종점 조회
- 구간 추가 시, 유효성 검사 추가
- 거리 wrapping

yunbaek committed:
feat: 구간 추가 기능 추가

- domain 레벨 구간 추가 기능 추가
- 구간 추가 관련 유닛 테스트 추가
- 구간 추가 실패 관련 기능 및 유닛 테스트 추가

yunbaek committed:
feat: 상행 종점 기준 역 정렬 기능 추가

yunbaek committed:
feat: 노선 기준 구간 추가 기능 구현

yunbaek committed:
test: 구간 추가 인수테스트 수정

yunbaek committed:
docs: 요구사항 목록 수정

shared-moon commented on src/main/java/nextstep/subway/domain/line/Section.java:
```diff
@@ -0,0 +1,104 @@
+		this.line = line;
+		this.upStation = upStation;
+		this.downStation = downStation;
+		this.distance = new Distance(distance);
```
`new Distance()` 에는 distance에 대한 유효성 검사가 같이 있는데, 도메인로직에 들어오기 전에 수행이 되면 어떨까요 ?

yunbaek replied:
`Distance` 에 대한 유효성 검사는 `Distance` 가 하면 될 것 같아서 그렇게 했는데, 피드백 주신 내용 생각해보니 `Distance` 나름 대로의 유효성을 가지고 있고,  `Section` 에서 생성할 때에 따로 유효성 검사가 있는 것도 좋아보입니다! 😄

shared-moon commented on src/main/java/nextstep/subway/application/line/LineService.java:
```diff
@@ -76,13 +77,27 @@ private LineResponse lineResponse(Line line) {
+		return lineResponse(line);
+	}
+
+	private Station upStation(Long upStationId) {
```
메서드명이 약간 find를 할거같은 느낌은 아닌 것 같아요!

yunbaek replied:
아 이 부분은 최근에 읽고 있는 `엘레강트 오브젝트` 란 책에서 
`빌더`는 명사로, `조정자`는 동사로 네이밍을 지어보라는 부분을 읽고
나름대로 적용해 보고 있었는데, 
자세히 생각해보니, 단순 빌더가 아니라 DB 에서 `검색`을 해와야 하기 때문에
네이밍을 수정하는게 좋아보입니다!
세심한 리뷰 정말 감사드립니다 😂  

[[엘레강트 오브젝트] 2-4장 메서드 이름을 신중하게 선택하세요](https://jessyt.tistory.com/141?category=967484)

제가 참고한 부분입니다!

shared-moon replied:
좋은 내용입니다 👍 약간 이번 경우와는 맞지 않는 것 같긴 한데, 매번 이렇게 의식적으로 신경쓰시는 것 좋아보여요 :)

shared-moon commented on src/main/java/nextstep/subway/domain/line/Section.java:
```diff
@@ -0,0 +1,104 @@
+		return station.equals(newStation);
+	}
+
+	@Override
```
보통 equals & hashcode 등의 오브젝트 기능을 오버라이드 한 경우에는 메서드를 맨 아래로 내려주는 편이에요!

shared-moon commented on src/main/java/nextstep/subway/domain/line/Section.java:
```diff
@@ -0,0 +1,104 @@
+		return Objects.hashCode(id);
+	}
+
+	public void updateUpStation(Station downStation, Integer distance) {
```
약간 update하는 경우가 구간 추가만 있는 건 아니라서 로직에 비해 이름이 너무 보편적인거같긴 한데, 마침 다음 미션이 구간삭제니까 그때 다시 고민해보죠!

shared-moon commented on src/main/java/nextstep/subway/domain/line/Sections.java:
```diff
@@ -0,0 +1,147 @@
+import nextstep.subway.exception.InvalidSectionAddException;
+
+@Embeddable
+public class Sections {
```
일급컬렉션 활용 좋습니다 👍

shared-moon commented on src/main/java/nextstep/subway/domain/line/Sections.java:
```diff
@@ -0,0 +1,147 @@
+		this.sections = sections;
+	}
+
+	public static Sections initialSections(Section section) {
```
이 초기화 메서드가 따로 필요한 이유가 있을까요 ?

yunbaek replied:
처음 노선을 생성할 때에, 기본이 되는 구간도 추가 된다는 개념으로 좀 더  명시적인 네이밍을 주고 싶어서 따로 선언해보았는데, 불필요할까요? 😅

shared-moon replied:
약간 밖에서 따로 불러줘야 하는 부분이라 initial 이라는 내용이 붙은게 조금 어색했어요! ( 재사용이 될 녀석도 아니라서 .. )
생성자나 of 등의 팩토리메서드로 방법을 더 제공해주는건 어떨까요 ?

shared-moon commented on src/main/java/nextstep/subway/domain/line/Sections.java:
```diff
@@ -0,0 +1,147 @@
+
+	public Station firstUpStation() {
+		List<Station> stations = allUpStations();
+		stations.removeAll(allDownStations());
```
`allUpStations()`에서 반환하는 List의 종류가 무엇이고, 현재 상태에서 `removeAll()` 수행 시 어떤 일이 일어날까요 ?

yunbaek replied:
`Collectors.toList` 는 `ArrayList` 를 반환하고, `ArrayList` 에서 `removeAll()` 을 하게 되면 내부적으로 `batchRemove` 메서드를 타게 되어, 인덱스도 다시 설정하는 등 비효율적으로 동작하는 것 같습니다.
그래서 `allUpStations`와 `allDownStations` 에서 리턴 되는 리스트를 `LinkedList` 로 변경하고 `removeIf` 메서드를 사용하여 O(n) 의 시간복잡도를 갖도록  수정했는데, 피드백에서 의도하신 부분이 맞는지는 확신이 안서네요  😂

shared-moon replied:
넵 ArrayList의 removeAll()은 그 동작 자체만으로도 굉장히 비효율적인 메서드라 그 부분을 말씀드리고 싶었어요!
추가로 처리방법에 대해서는 remove보단 탐색을 수행하는게 낫지 않을까 하는 의견이긴 했습니다!

shared-moon commented on src/main/java/nextstep/subway/domain/line/Sections.java:
```diff
@@ -0,0 +1,147 @@
+	}
+
+	private void rearrange(Section section) {
+		if (sameUpStation(section.getUpStation())) {
```
이런 작업들은 혹시 어플리케이션 레벨이 아니라 쿼리를 통해 처리 해 볼 수 있지 않을까요 ?

yunbaek replied:
그렇지 않아도, 컬렉션 조회를 너무 많이 하는 것 같아 찝찝했는데 바로 찝어주시네요. 😂 
팁 주신대로 query 를 통해 조회 후에 재배치 할 수 있도록 리팩토링 해봤습니다!

shared-moon commented on src/main/java/nextstep/subway/domain/line/vo/Distance.java:
```diff
@@ -0,0 +1,55 @@
+
+@Embeddable
+@Access(AccessType.FIELD)
+public class Distance {
```
VO 사용 좋습니다 👍

shared-moon commented on src/main/java/nextstep/subway/exception/InvalidSectionAddException.java:
```diff
@@ -0,0 +1,8 @@
+package nextstep.subway.exception;
+
+public class InvalidSectionAddException extends IllegalArgumentException {
```
상속은 왠만하면 RuntimeException을 받는게 좋을 것 같아요!

yunbaek replied:
말씀하신대로 `unchecked exception` 에 대해서는 `RuntimeException` 을 상속하는 것이 좋아보이네요!! 👍

shared-moon reviewed (changes requested):
경윤님 안녕하세요!
3단계 미션 잘 진행 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데, 한번 정리하고 가면 좋을 것 같아서 확인 후 리뷰 재요청 부탁드릴게요!
화이팅입니다 🔥

Commit:
refactor: 피드백 반영

- UncheckedException 예외 상속 변경(IllegalArgumentException -> RuntimeException)
- 재정의 메서드(equals & hashcode) 위치 변경
- 거리 유효성 검증 추가
- 메서드 네이밍 변경
- 리스트 중복 제거 로직 변경

Commit:
refactor: 구간 재배치 로직 리팩토링

- 재배치할 구간을 query 를 통해 조회
- 재배치 필요시에만 재배치

Commit:
chore: override 메서드 위치 변경

Commit:
chore: 오타수정

shared-moon commented on src/main/java/nextstep/subway/application/line/SectionService.java:
```diff
@@ -0,0 +1,25 @@
+	}
+
+	@Transactional(readOnly = true)
+	public List<Section> findSectionsToUpdate(Station upStation, Station downStation) {
```
👍

shared-moon commented on src/main/java/nextstep/subway/domain/line/vo/Distance.java:
```diff
@@ -48,8 +48,8 @@ public int hashCode() {
-	public Distance subtract(Integer distance) {
-		return new Distance(this.value - distance);
+	public Distance subtract(Distance distance) {
+		return new Distance(this.value - distance.value);
```
불변객체 활용 좋네요 👍

shared-moon commented on src/main/java/nextstep/subway/domain/line/Sections.java:
```diff
@@ -66,48 +66,29 @@ private Section sectionByUpStation(Station station) {
 	public Station firstUpStation() {
 		List<Station> stations = allUpStations();
-		stations.removeAll(allDownStations());
+		stations.removeIf(station -> allDownStations().contains(station));
```
이 종점 찾는 부분을 약간 논리로 풀어보자면
```
- upStation이 요청받은 upStation과 같은 경우 ( 중간삽입 )
- upStation이 요청받은 downStation과 같은 경우 ( 상행 종점 연장 )
- downStation이 요청받은 downStation과 같은 경우 ( 중간삽입 )
- downStation이 요청받은 upStation과 같은 경우 ( 하행 종점 연장 )
```
이런식의 조건들이 나올 것 같아요! 이 내용을 적용해보면 저희가 `구간 연장` 이라는 행위를 추상화 해 볼 수 있지 않을까요 ?

shared-moon reviewed (approved):
경윤님 안녕하세요!
피드백 반영 잘 진행 해 주셨네요 :)
코멘트를 몇 개 남겨 두었는데, 다음 미션 진행 전에 확인 부탁드려요!
고생 많으셨습니다 :)
